### PR TITLE
Pulling firebase utilities

### PR DESCRIPTION
**Pull request by Buchimi**
I was able to create tools and connect our backend code to firebase servers. 
If you have access to the firebase project. You need to create an .env.local file to be able to use firebase and an .env file to run the firebase tests.
Reach out to me on instructions.

install the firebase package and mocha
``` 
npm i 
```

### CODE REVIEW

**Comment by Buchimi** [comment]
Looks like i need to do a git pull and get the base branch

**Comment by Jake3231** [comment]
Looks great! It's exciting to see the initial Firebase integration working so well!

File: backend_tools/db_provider.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        }
+        catch (error) {
+            console.error(`failure saving user, got error: ${error}`);
+            return false;
+        }
+    }
+
+    async createClub(club: Club): Promise<boolean> {
+        try {
+            //const clubCollection = collection(this.db, this.club_path)
```
Comment (Jake3231) [inline backend_tools/db_provider.ts:30]: Is this line needed currently?

File: backend_tools/db_provider.ts [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        return ref.data() as unknown as User;
+    }
+
+    async getClub(club_name: string): Promise<Club> {
+        const clubReference = doc(this.db, this.club_path, club_name)
+        const ref = await getDoc<DocumentData>(clubReference)
+        return ref.data() as Club;
+    }
+
+    async getSomeClubs(): Promise<Club[] | null> {
```
Comment (Jake3231) [inline backend_tools/db_provider.ts:55]: Could this potentially be renamed to best indicate what it is doing, if I'm understanding correctly? 
```suggestion
    async getAllClubs(): Promise<Club[] | null> {
```

File: backend_tools/picture_uploader.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import { deleteObject, getBlob, getStorage, ref, uploadBytes, UploadResult } from "firebase/storage"
+class PictureUploader {
+    storage = getStorage()
+    image_path = "images"
+
+    /// Uploads an image to our storage bucket
+    /// if image name is undefined, it
```
Comment (Jake3231) [inline backend_tools/picture_uploader.ts:7]: Is this comment fully completed?

File: backend_tools/test_db_provider.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import DbProvider from "./db_provider";
+
+class DbTestProvider extends DbProvider {
+    user_path: string = "TestUsers"
```
Comment (Jake3231) [inline backend_tools/test_db_provider.ts:4]: Should these potentially be changed to be production paths?
```suggestion
    user_path: string = "Users"
```

File: models/club.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import { DocumentReference, DocumentData } from "firebase/firestore"
+import Contacts from "./contacts"
+
+interface Club {
+    name: string
+    events?: DocumentReference<DocumentData>[] //references to other events,
```
Comment (Jake3231) [inline models/club.ts:6]: Is this comment partially cut off? It may have not come through correctly.

File: models/contacts.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
+interface Contacts {
```
Comment (Jake3231) [inline models/contacts.ts:1]: This seems great to define a structure, though would we want to implement a solution that could be more general as to account for contact methods we didn't think of?

File: models/event.ts [hunk 7]
```diff
@@ -0,0 +1,8 @@
+interface Event {
+    name: string,
+    hostclub: string    //club name. THIS NEEDS TO BE THE SAME AS THE CLUB ID IN THE DATABASE.
```
Comment (Jake3231) [inline models/event.ts:3]: Per the comment, does anything need to change here prior to merging?

File: tests/test.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+import { config } from "dotenv"
+import Firebase from "../backend_tools/firebase"
+import User from "../models/user"
+import Year from "../models/year"
+import Role from "../models/role"
+import Career from "../models/career"
+import DbTestProvider from "../backend_tools/test_db_provider"
+
+
+// const firebaseConfig = {
```
Comment (Jake3231) [inline tests/test.ts:13]: If the commented code is completely unneeded, it may be good to remove it

**Review by Jake3231 — CHANGES_REQUESTED** [review]
Overall looks great, just left a few specific questions.

File: backend_tools/db_provider.ts (hunk 1 above)
Comment (Buchimi) [reply backend_tools/db_provider.ts:30]: No it's not. will be deleted

File: backend_tools/db_provider.ts (hunk 2 above)
Comment (Buchimi) [reply backend_tools/db_provider.ts:55]: You're right

File: backend_tools/picture_uploader.ts (hunk 3 above)
Comment (Buchimi) [reply backend_tools/picture_uploader.ts:7]: No, i completed it

File: backend_tools/test_db_provider.ts (hunk 4 above)
Comment (Buchimi) [reply backend_tools/test_db_provider.ts:4]: No they shouldn't
These paths are for testing purposes only.

Our test scripts will make use of the db provider for testing

File: models/event.ts (hunk 7 above)
Comment (Buchimi) [reply models/event.ts:3]: We can remove the comment. It looks like the club name is registered as it's ID.

File: models/contacts.ts (hunk 6 above)
Comment (Buchimi) [reply models/contacts.ts:1]: I guess we could have an opportunity to upload a link?

File: models/club.ts (hunk 5 above)
Comment (Buchimi) [reply models/club.ts:6]: I may need to revise the club datastructure first

File: models/event.ts (hunk 7 above)
Comment (Buchimi) [reply models/event.ts:3]: Sorry, this is an event. Not a club. 
I don't think we should make use of functions that work with events yet

File: tests/test.ts (hunk 8 above)
Comment (Buchimi) [reply tests/test.ts:13]: i removed it

**Comment by Buchimi** [comment]
I have made the requested changes

**Review by Jake3231 — COMMENTED** [review]
Overall looks good!

File: pages/index.tsx [hunk 9]
```diff
@@ -2,10 +2,9 @@ import Head from 'next/head';
 import Image from 'next/image';
 import EventCard from '../components/EventCard';
 import banner from '../public/banner.png';
-
```
Comment (Jake3231) [inline pages/index.tsx:5]: Is it expected that this space is removed?

**Review by Jake3231 — APPROVED** [review]

File: pages/index.tsx (hunk 9 above)
Comment (Buchimi) [reply pages/index.tsx:5]: It's just a space, shouldn't matter

File: models/club.ts [hunk 10]
```diff
@@ -0,0 +1,12 @@
+import { DocumentReference, DocumentData } from "firebase/firestore"
+import Contacts from "./contacts"
+import Event from "./event"
+
+interface Club {
```
Comment (ZzRanger) [inline models/club.ts:5]: Just a suggestion...maybe put all your models in the same file or colocate your models with your query methods. A lot of these model files are very small so might be easier to condense it in a single place.

**Review by ZzRanger — APPROVED** [review]
Overall looks good! From my experience I've noticed that I end up changing my data models and query methods a lot when building out an MVP. Consider potentially finding ways to colocate information so that the burden of changing the models or the query methods is less taxing.

File: backend_tools/db_provider.ts [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+      const userCollection = collection(this.db, this.user_path);
+      await addDoc(userCollection, user);
+      return true;
+    } catch (error) {
+      console.error(`failure saving user, got error: ${error}`);
+      return false;
+    }
+  }
+
+  async createClub(club: Club): Promise<boolean> {
```
Comment (ZzRanger) [inline backend_tools/db_provider.ts:41]: Maybe instead of returning true or false return the club id | undefined

Comment (ZzRanger) [reply backend_tools/db_provider.ts:41]: Same thing w/ user

File: backend_tools/db_provider.ts [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+      return true;
+    } catch (error) {
+      console.error(`failure saving club, got error: ${error}`);
+      return false;
+    }
+  }
+
+  async getUser(userid: string): Promise<User> {
+    const userReference = doc(this.db, this.user_path, userid);
+    const ref = await getDoc<DocumentData>(userReference);
```
Comment (ZzRanger) [inline backend_tools/db_provider.ts:56]: How does firebase handle if a resource doesn't exist given a certain query? Might be nice to potentially add some error state handling for that here

File: backend_tools/db_provider.ts (hunk 11 above)
Comment (Buchimi) [reply backend_tools/db_provider.ts:41]: The club id for this function is the name of the club

File: backend_tools/db_provider.ts (hunk 12 above)
Comment (Buchimi) [reply backend_tools/db_provider.ts:56]: returns undefined

File: models/club.ts (hunk 10 above)
Comment (Buchimi) [reply models/club.ts:5]: i see, i prefer them as multiple files. don't they get optimized away by webpack or some similar technology?

**Review by RubenOlano — APPROVED** [review]
I think this looks good, I think we can maybe use a parser library to handle some of the parsings whenever passing data to frontend, but we can look at that in the future

**Review by Jake3231 — APPROVED** [review]
These changes look good!

**Pull request merged**